packages/fxa-auth-server/lib/payments/stripe.js

```diff
@@ -1015,6 +1011,179 @@ class StripeHelper {
+      planAmountNew,
+      closeDate: event.created,
+    };
+
+    if (!cancelAtPeriodEndOld && cancelAtPeriodEndNew && !planOld) {
+      return this.extractSubscriptionUpdateCancellationDetailsForEmail(
+        subscription,
+        baseDetails
+      );
+    } else if (cancelAtPeriodEndOld && !cancelAtPeriodEndNew && !planOld) {
+      return this.extractSubscriptionUpdateReactivationDetailsForEmail(
+        subscription,
+        baseDetails
+      );
+    } else if (!cancelAtPeriodEndOld && !cancelAtPeriodEndNew && planOld) {
+      return this.extractSubscriptionUpdateUpgradeDowngradeDetailsForEmail(
+        subscription,
+        baseDetails,
+        customer,
+        productOrderNew,
+        planOld
+      );
+    }
+
+    // unknown update scenario, but let's return some details anyway
+    return baseDetails;
+  }
+
+  async extractSubscriptionUpdateCancellationDetailsForEmail(
+    subscription,
+    baseDetails
+  ) {
+    // https://github.com/mozilla/subhub/blob/e224feddcdcbafaf0f3cd7d52691d29d94157de5/src/hub/vendor/customer.py#L598
+    return {
+      ...baseDetails,
+      isCancellation: true,
+      cancelledAt: subscription.canceled_at,
+      cancelAt: subscription.cancel_at,
+      cancelAtPeriodEnd: subscription.cancel_at_period_end,
+      currentPeriodStart: subscription.current_period_start,
+      currentPeriodEnd: subscription.current_period_end,
+      invoiceId: subscription.latest_invoice,
+    };
+  }
+
+  async extractSubscriptionUpdateReactivationDetailsForEmail(
+    subscription,
+    baseDetails
+  ) {
+    // https://github.com/mozilla/subhub/blob/e224feddcdcbafaf0f3cd7d52691d29d94157de5/src/hub/vendor/customer.py#L625
+    const invoice = await this.expandResource(
+      subscription.latest_invoice,
+      INVOICES_RESOURCE
+    );
+    const charge = await this.expandResource(invoice.charge, CHARGES_RESOURCE);
+    const { brand, last4 } = charge.payment_method_details.card;
+    return {
+      ...baseDetails,
+      isReactivation: true,
+      invoiceId: subscription.latest_invoice,
+      currentPeriodEnd: subscription.current_period_end,
+      brand,
+      last4,
+    };
+  }
+
+  async extractSubscriptionUpdateUpgradeDowngradeDetailsForEmail(
+    subscription,
+    baseDetails,
+    customer,
+    productOrderNew,
+    planOld
+  ) {
+    // https://github.com/mozilla/subhub/blob/e224feddcdcbafaf0f3cd7d52691d29d94157de5/src/hub/vendor/customer.py#L643
+    const abbrevProductOld = await this.expandAbbrevProductForPlan(planOld);
+    const { interval: productPaymentCycleOld, amount: planAmountOld } = planOld;
+    const {
+      product_id: productIdOld,
+      product_name: productNameOld,
+    } = abbrevProductOld;
+    const {
+      productOrder: productOrderOld,
+      emailIconURL: productIconURLOld = '',
+      downloadURL: productDownloadURLOld = '',
+    } = this.mergeMetadata(planOld, abbrevProductOld);
+
+    let isUpgrade = false;
+    let isDowngrade = false;
```

Comment:
Do we really need to have two booleans for this? I would imagine that something like this might be more useful:
```suggestion
    if (productOrderNew === productOrderOld) {
     throw new SubscriptionUpdateNotProductChageError(); // something to indicate that the change made to the subscription does not meet the criteria for an upgrade/downgrade email
     }
     
     // Now we can say that if isUpgrade is false, we have a downgrade, since the invalid case has thrown an exception
     let isUpgrade = false;
     if (productOrderNew > productOrderOld) {
         isUpgrade = truw;
     }
```

Reply:
There are actually four booleans in total, for each of the emails we potentially send on subscription update: https://github.com/mozilla/fxa/pull/4949/files#diff-eedc9e44d94c523680530e9c145c9496R903

I could switch that to a single multi-state flag, but I figured the booleans were more clear.

Reply:
You know what might work here, is instead of having these booleans, we do something like
```
const UPGRADE_EVENT = 'upgrade';
const DOWNGRADE_EVENT = 'downgrade';
const REACTIVATION_EVENT = 'reactivation';
const CANCELLATION_EVENT = 'cancellation';

....

    let updateEventType = DOWNGRADE_EVENT;
    if (productOrderNew > productOrderOld) {
         updateEventType = UPGRADE_EVENT;
     }

....
switch(eventDetails.updateEventType) {
   case UPGRADE_EVENT: ...
}

Reply:
I think something like that would make things read in a very clear manner